bokehjs/src/lib/api/plotting.ts

```diff
@@ -630,30 +630,47 @@ export class Figure extends Plot {
     return this._marker(models.Scatter, args)
   }
 
-  _pop_colors_and_alpha(cls: Class<HasProps>, attrs: Attrs, prefix: string = "",
-                        default_color: Color = _default_color, default_alpha: number = _default_alpha): Attrs {
-    const result: Attrs = {}
+  _pop_visuals(cls: Class<HasProps>, props: Attrs, prefix: string = "",
+                        defaults: Attrs = {}, trait_defaults: Attrs = {}): Attrs {
 
-    const color = _with_default(attrs[prefix + "color"] as any, default_color)
-    const alpha = _with_default(attrs[prefix + "alpha"] as any, default_alpha)
+    const _split_feature_trait = function(ft: string): string {
+      ft = ft.split('_', 2)
+      if (ft.length==2) { return ft }
+      else { return ft.concat([null]) }
+    }
+    const _is_visual = function(ft: string): boolean {
+      [feature, trait] = _split_feature_trait(ft)
+      return (feature in ['line', 'fill', 'text']) && (trait != null)
```

Comment:
> `feature in ['line', 'fill', 'text']`

Unfortunately this isn't Python, so the meaning of this code is pretty surprising. Let me explain. An array in JS can be viewed as a key-value map (or an object), where keys are integer indices and values are actual values in an array. `in` operator works on objects and checks if a key exists in an object. So you can ask if an index exists in an array (not particularly useful for dense arrays, but sparse arrays are perfectly possible), so you can ask if e.g. `0 in ["a value"]`, which is true, but asking if `"a value" in ["a value"]` doesn't make sense, be cause Array object doesn't have `a value` property. The correct way is to write `["a value"].includes("a value")`. TypeScript doesn't complain about this, because despite the code not making sense in the way we would expect it to work, is still perfectly valid JavaScript and has merit in some contexts. At some point we will have ESLint rules that would warn about this kind of nonsense code. This also makes efficient arrays sometimes tricky, because it's easy to turn a dense array into a sparse one, which disables any JS engine's optimizations, making operations such arrays slow.

Reply:
Thanks, should be fixed now. Been a while since I last used JS, that one escaped me. :)